test/unit/core/algorithm/deferred_config_element_base_test.cpp

```diff
@@ -0,0 +1,146 @@
+// ============================================================================
+//                 SeqAn - The Library for Sequence Analysis
+// ============================================================================
+//
+// Copyright (c) 2006-2018, Knut Reinert & Freie Universitaet Berlin
+// Copyright (c) 2016-2018, Knut Reinert & MPI Molekulare Genetik
+// All rights reserved.
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are met:
+//
+//     * Redistributions of source code must retain the above copyright
+//       notice, this list of conditions and the following disclaimer.
+//     * Redistributions in binary form must reproduce the above copyright
+//       notice, this list of conditions and the following disclaimer in the
+//       documentation and/or other materials provided with the distribution.
+//     * Neither the name of Knut Reinert or the FU Berlin nor the names of
+//       its contributors may be used to endorse or promote products derived
+//       from this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL KNUT REINERT OR THE FU BERLIN BE LIABLE
+// FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+// DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+// SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+// CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+// LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+// OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
+// DAMAGE.
+//
+// ============================================================================
+
+#include <gtest/gtest.h>
+
+#include <meta/meta.hpp>
+
+#include <seqan3/core/algorithm/deferred_config_element_base.hpp>
+#include <seqan3/core/algorithm/configuration.hpp>
+
+using namespace seqan3;
+
+template <int _d>
+class bar_static : public detail::config_element_base<bar_static<_d>>
+{
+    friend class detail::config_element_access<bar_static<_d>>;
+
+    int state{_d};
+};
+
+class bar : public detail::deferred_config_element_base<bar>
+{
+    friend class detail::config_element_access<bar>;
+
+    template <typename fn_t, typename configuration_t>
+    constexpr auto invoke(fn_t && fn, configuration_t && config) const
+    {
+        using new_cfg_t = detail::replace_config_with_t<std::remove_reference_t<configuration_t>,
+                                                        bar,
+                                                        bar_static<1>>;
+        return fn(new_cfg_t{config});
+    }
```

Comment:
I'm a bit confused. The example above was not working so I don't know if you wanted to have here multiple static types depending on some runtime parameters?